src/engine/openmp.cc

```diff
@@ -74,17 +74,17 @@ void OpenMP::set_reserve_cores(int cores) {
   reserve_cores_ = cores;
 #ifdef _OPENMP
   if (reserve_cores_ >= omp_thread_max_) {
-    omp_set_num_threads(1);
+    omp_thread_max_ = 1;
   } else {
-    omp_set_num_threads(omp_thread_max_ - reserve_cores_);
+    omp_thread_max_ -= reserve_cores_;
```

Comment:
I think this does not solve multithreading problem but rather causing race condition.

Reply:
isn't this atomic?

Reply:
Say at one time reserve_cores_ = 4 and omp_thread_max_ = 8

Thread 1: checks `reserve_cores_ >= omp_thread_max` is false so goto the else branch; at this time, thread2 is executing in the core and reserve_cores = 12 and therefore will update the omp_thread_max to 1. And when it comes to Thread 1 to run, it will execute `omp_thread_max_ -= reserve_cores_`. Now `omp_thread_max_` will be negative.